### PR TITLE
Bootstrap upgrade 3to4

### PR DESCRIPTION
**Pull request by hq-ods**
Issue #5422 

### Proposed workplan:

- [X] Backup base and template directory that uses bootstrap 3
- [X] Update bootstrap build parameters
- [X] Remove deprecated bootstrap 3 classes to their bootstrap 4 counterparts
   - [X] Headers
   - [X] Breadcrumbs
   - [X] Navigation
   - [X] Facet Navigation
   - [X] Forms
   - [X] Media grid (to be replaced by BS4 cards system)
   - [X] Pagination
   - [X] Badges
   - [X] Buttons

- [X] Migrate all LESS to SCSS preprocessing
- [X] Update all templates
- [X] Update applicable python scripts
- [ ] Update unit tests
- [ ] Update documentation
- [ ] Write migration guide for BS3 to 4

### Features:

- [ ] includes tests covering changes
- [ ] includes updated documentation
- [X] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

Please [X] all the boxes above that apply

### CODE REVIEW

**Comment by Zharktas** [comment]
@hq-ods bootstrap itself should be installed from npm via [package.json](https://github.com/ckan/ckan/blob/master/package.json#L7), there's a [gulp-script](https://github.com/ckan/ckan/blob/master/gulpfile.js#L37..L43) which tries to copy the files in their respectable places for allowing somewhat easy updates, instead of just manually adding them to version control. These should be updated too.

**Comment by hq-ods** [comment]
@Zharktas thanks for pointing that out. Updated in the latest commit. I'm fairly new to this project so please feel free to provide any comments whenever.

Also, I noticed that right-to-left bootstrap-rtl.less file is a 3rd party. Since there are no official implementation to address it in v4, I found another library that looks like it's [actively supported ](https://github.com/PerseusTheGreat/bootstrap-4-rtl).

**Comment by Zharktas** [comment]
@hq-ods as https://github.com/ckan/ckan/pull/5360 got merged, you should sync with master as templates and less files were changed.

**Comment by hq-ods** [comment]
@Zharktas some of the gulp-sass dependencies are failing the test_coding_standards, can you advise on how these should be handled? 

image

File: ckan/lib/helpers.py [hunk 1]
```diff
@@ -1094,7 +1095,7 @@ def build_extra_admin_nav():
 def _make_menu_item(menu_item, title, **kw):
     ''' build a navigation item used for example breadcrumbs
 
-    outputs <li><a href="..."></i> title</a></li>
+    outputs <li class="nav-item nav-link"><a href="..."></i> title</a></li>
```
Comment (boykoc) [inline ckan/lib/helpers.py:1098]: although a private function so not in the docs I believe, this output isn't shown as code. Could be updated while making this update.

outputs ``<li class="nav-item nav-link"><a href="..."></i> title</a></li>``

File: ckan/lib/helpers.py [hunk 2]
```diff
@@ -1020,7 +1020,8 @@ def build_nav_main(*args):
 def build_nav_icon(menu_item, title, **kw):
     '''Build a navigation item used for example in ``user/read_base.html``.
 
-    Outputs ``<li><a href="..."><i class="icon.."></i> title</a></li>``.
+    Outputs ``<li class="nav-item nav-link"><a href="..."><i class="icon..">
```
Comment (boykoc) [inline ckan/lib/helpers.py:1023]: Inline code operator used for multi-line code output. I think this is one of the test failures.

**Comment by Zharktas** [comment]
I know this hasn't had much advances in a few months but now that Bootstrap 5 is in [beta](https://blog.getbootstrap.com/2020/12/07/bootstrap-5-beta-1/) should we just upgrade to 5 directly ?

**Comment by tino097** [comment]
Any progress on this one?

Do we need to have similar approach as the #3547 ?

**Comment by hq-ods** [comment]
Might have to, unfortunately i got pulled into other things at work that's taking all my time. If someone else want to take a look at this or maybe even just jump from 4 to 5 as Zharktas mentioned.

**Comment by Zharktas** [comment]
This is superseded by bs5 upgrade, so closing this.

**Pull request closed without merging**